libraries/YarpPlugins/CanBusHico/CanBusHico.cpp

```diff
@@ -2,74 +2,114 @@
 
 #include "CanBusHico.hpp"
 
+#include <unistd.h>
+#include <assert.h>
+#include <stdint.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/select.h>
+#include <sys/time.h>
+
+#include <cstring>
+
+#include <ColorDebug.hpp>
+
 // -----------------------------------------------------------------------------
 
-bool roboticslab::CanBusHico::sendRaw(uint32_t id, uint16_t len, uint8_t * msgData)
+bool roboticslab::CanBusHico::setFdMode(bool requestedBlocking)
 {
+    bool currentlyBlocking = (fcntlFlags & O_NONBLOCK) != O_NONBLOCK;
 
-    struct can_msg msg;
-    memset(&msg,0,sizeof(struct can_msg));
-    msg.ff = FF_NORMAL;
-    msg.id = id;
-    msg.dlc = len;
-    memcpy(msg.data, msgData, len*sizeof(uint8_t));
-
-    canBusReady.wait();
-    if ( write(fileDescriptor,&msg,sizeof(struct can_msg)) == -1)
+    if (currentlyBlocking != requestedBlocking)
     {
-        canBusReady.post();
-        CD_ERROR("%s.\n", strerror(errno))
-        return false;
+        int flag = requestedBlocking ? ~O_NONBLOCK : O_NONBLOCK;
+        return ::fcntl(fileDescriptor, F_SETFL, fcntlFlags & flag) != -1;
     }
-    canBusReady.post();
 
     return true;
 }
 
 // -----------------------------------------------------------------------------
 
-int roboticslab::CanBusHico::read_timeout(struct can_msg *buf, unsigned int timeout)
+bool roboticslab::CanBusHico::setDelay()
 {
-
     fd_set fds;
     struct timeval tv;
-    int sec,ret;
-    FD_ZERO(&fds);
 
-    sec=timeout/1000;
-    tv.tv_sec=sec;
-    tv.tv_usec=(timeout-(sec*1000))*1000;
+    tv.tv_sec = DELAY;
+    tv.tv_usec = DELAY * 1000 * 1000;
 
-    FD_SET(fileDescriptor,&fds);
+    FD_ZERO(&fds);
+    FD_SET(fileDescriptor, &fds);
 
     //-- select() returns the number of ready descriptors, or -1 for errors.
-    ret=select(fileDescriptor+1,&fds,0,0,&tv);
-    if(ret==0)
+    int ret = ::select(fileDescriptor + 1, &fds, 0, 0, &tv);
+
+    if (ret <= 0)
     {
-        //-- Commenting select() timeout as way too verbose, happens all the time.
-        //CD_DEBUG("select() timeout.\n");
-        return 0;  // Return 0 on select timeout.
+        //-- No CD as select() timeout is way too verbose, happens all the time.
+        // Return 0 on select timeout, <0 on select error.
+        return false;
```

Comment:
This is totally wrong. `CanBusHico::setDelay` must be repurposed to reflect one of three possible states:

* error (`ret < 0`)
* timeout (`ret == 0`), don't call `::read` afterwards
* success (`ret > 0`), tell callers to issue a `::read` operation

Also, I have to take a closer looks at every single call to `canRead` (perhaps `canWrite`, too) and ensure that it makes sense.